### ISSUE TITLE
Text is blurry & jitters left and right

### ISSUE THREAD

**Issue by max-vogler** [excaliburjs/Excalibur#2674]
<img width="127" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/864168/35650d33-7480-43a0-947f-f1c54a6b61f1">

### Steps to Reproduce

I'm writing a debug information screen element, that updates every frame to show the rough system load.

```ts
export class DebugInfo extends ScreenElement {
  private readonly text = new Text({
    font: new Font({
      family: "Gorgeous Pixel",
      filtering: ImageFiltering.Pixel,
      smoothing: false,
      size: 8, // trying sizes 5-20 does not fix the issue.
    }),
    text: "",
  });

  onInitialize(_engine: Engine): void {
    this.graphics.use(this.text);
  }

  onPreUpdate(engine: Engine, _delta: number): void {
    if (engine.maxFps) {
      const budget = 1000 / engine.maxFps;
      const load = engine.stats.prevFrame.duration.total / budget;
      const pct = Math.round(load * 100);
      this.text.text = `Load: ${pct} %`;
    } else {
      const fps = Math.round(engine.stats.prevFrame.fps);
      this.text.text = `FPS: ${fps}`;
    }
  }
}
```

I added it to my game that upscales pixels 3x without smoothing:

```ts
class Game extends Engine {
  constructor() {
    super({
      width: 450,
      height: 450,
      resolution: { // removing upscaling does not fix the issue
        height: 150,
        width: 150,
      },
      antialiasing: false,
      suppressHiDPIScaling: true,
    });
  }

  initialize() {
    this.currentScene.add(new DebugInfo({ x: 0, y: 10 }));
  }
```

And loaded the font family in CSS:

```css
@font-face {
  font-family: "Gorgeous Pixel";
  src: url("/Gorgeous Pixel.ttf");
}
```

### Expected Result

See the [Georgeous Pixel font page](https://www.dafont.com/gorgeous-pixel.font?text=Load%3A+8+%25&back=bitmap) for a demo of how the font should look.

### Actual Result

The font is scaled in a blurry way, despite antialiasing and smoothing disabled. Additionally, the text wildly jumps left and right when the text changes (which changes on most frames).

### Environment

- browsers and versions: Chrome 114
- operating system: macOS 13
- Excalibur versions: 0.27.0

### Current Workaround

None.

**Comment by eonarheim** [comment]
@max-vogler Thanks for the issue!

I'll look into this more closely to see what's up, it's very possible there is some regression in text.

Off-hand I know text can be tricky at lower resolutions because it needs to render the text onto a canvas with limited pixels. In this case 150x150, the way excalibur renders these browser fonts is by rasterizing them into an offscreen 2D canvas and using that as an image to render on the screen. You can lose quality because of sampling when you need to downscale images like that which can potentially cause some of the artifacts.

One possible workaround to get more "crisp" text is to bump the resolution of the game up by 3-4x (giving you more pixels to work with and avoid downsampling on the text) and artificially scale up all your graphics to keep the keep the low-res feel.

Another strategy is to use a `SpriteFont` to render the text, which wont suffer as badly at lower resolutions.

**Comment by eonarheim** [comment]
The flickering left/right is wild, not sure why it's doing that yet. I'm digging in deeper. 

My guess is there is something funky about how Excalibur is computing the bounds for the text and positioning it in an odd way.

**Comment by eonarheim** [comment]
@max-vogler I have a couple work arounds for you!

So for the text left/right shift it does appear that something is odd with the bounds calculation so excalibur is trying to position the text as the bounds change. Manually interrupting that and setting static bounds fixes it for now (I'll dig into the root cause)

```typescript
import { SCALE } from "./config";

export class DebugInfo extends ScreenElement {
    private readonly text = new Text({
      font: new Font({
        family: "Gorgeous Pixel",
        filtering: ImageFiltering.Pixel,
        smoothing: false,
        size: 8 * SCALE, // trying sizes 5-20 does not fix the issue.
      }),
      text: "",
    });
  
    onInitialize(_engine: Engine): void {
      this.graphics.use(this.text);
    }
  
    onPreUpdate(engine: Engine, _delta: number): void {
      if (engine.maxFps) {
        const budget = 1000 / engine.maxFps;
        const load = engine.stats.prevFrame.duration.total / budget;
        const pct = Math.round(load * 100);
        this.text.text = `Load: ${pct} %`;
        console.log(this.text.localBounds.width);
      } else {
        const fps = Math.round(engine.stats.prevFrame.fps);
        this.text.text = `FPS: ${fps}`;
      }
      // manually set the graphics component's bounds 
      this.graphics.localBounds = BoundingBox.fromDimension(100, 20, Vector.Zero);
    }
  }
```  

As for the blurry text, upscaling the resolution seems to work around that issue as well.

```typescript
// config.ts
export const SCALE = 4;
```
```typescript
class Game extends Engine {
    constructor() {
      super({
        width: 450,
        height: 450,
        resolution: { // removing upscaling does not fix the issue
          height: 150 * SCALE,
          width: 150 * SCALE,
        },
        antialiasing: false,
        suppressHiDPIScaling: true,
        maxFps: 60
      });
    }
  
    initialize() {
      this.currentScene.add(new DebugInfo({ x: 0, y: 10 * SCALE }));
    }
}
```
![text](https://github.com/excaliburjs/Excalibur/assets/612071/e2524358-0299-4387-bac3-eb0e925d497a)

**Comment by eonarheim** [comment]
@max-vogler Quick update! I've started digging into the weird jittering alignment it looks like it's been fixed in the most recent alpha so that bounds workaround is not necessary `0.28.0-alpha.687`

I think it was fixed here when we overhauled text and fixed other alignment bugs https://github.com/excaliburjs/Excalibur/commit/730d621805d32104c18dc7ca8d9e97c7f212bb2b

**Comment by max-vogler** [comment]
Thanks for your in depth replies! I can confirm that the horizontal jittering is fixed in `0.28.0-alpha.687`.

For the blurriness, I still see faint blurriness in your sample video. Going more the SpriteFont route would be acceptable as a workaround for me, e.g. do you think me providing a hardcoded pre-rendering size could be a workaround?

**Comment by eonarheim** [comment]
Hi @max-vogler, this thread fell off my radar!

I've poked around a bit and built something that I think looks pretty good. I can confirm that you don't need to do any scaling hacks to have the pixel aesthetic when using `SpriteFont`s.

If by hardcoding you mean providing a sprite font that is exactly the pixel size in the resolution you want I think that it will totally work around the issue.

![spritefont](https://github.com/excaliburjs/Excalibur/assets/612071/38b9a204-be56-4c63-97b8-45442609dcea)

I made a quick sprite font `spritefont.png` 
![spritefont](https://github.com/excaliburjs/Excalibur/assets/612071/044ef99a-872a-47bc-96f5-fc4badd36725)


```typescript
var SCALE = 1;
var image = new ex.ImageSource('./icon@2x.png');
var spriteFontImage = new ex.ImageSource('./spritefont.png', false, ex.ImageFiltering.Pixel);
var spriteSheet = ex.SpriteSheet.fromImageSource({
  image: spriteFontImage,
  grid: {
    rows: 3,
    columns: 16,
    spriteWidth: 16,
    spriteHeight: 16,
  },
});

var spriteFont = new ex.SpriteFont({
  alphabet: '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ,!\'&."?-:% ',
  caseInsensitive: true,
  spriteSheet: spriteSheet,
  spacing: -8,
});

class DebugInfo extends ex.ScreenElement {
  private readonly text = new ex.Text({
    font: new ex.Font({
      family: "Gorgeous Pixel",
      filtering: ex.ImageFiltering.Pixel,
      smoothing: false,
      size: 8 * SCALE, // trying sizes 5-20 does not fix the issue.
    }),
    text: "",
  });

  private readonly spriteFontText = new ex.Text({
    text: "",
    font: spriteFont
  });

  onInitialize(_engine: ex.Engine): void {
    // this.graphics.show(this.text);
    this.graphics.use(this.spriteFontText);
  }

  onPreUpdate(engine: ex.Engine, _delta: number): void {
    if (engine.maxFps) {
      const budget = 1000 / engine.maxFps;
      const load = engine.stats.prevFrame.duration.total / budget;
      const pct = Math.round(load * 100);
      this.text.text = `Load: ${pct} %`;
      this.spriteFontText.text = `Load: ${pct}%`;
    } else {
      const fps = Math.round(engine.stats.prevFrame.fps);
      this.text.text = `FPS: ${fps}`;
      this.spriteFontText.text = `FPS: ${fps}`;
    }
  }
}

class Game4 extends ex.Engine {
  constructor() {
    super({
      width: 450,
      height: 450,
      resolution: { // removing upscaling does not fix the issue
        height: 150 * SCALE,
        width: 150 * SCALE,
      },
      antialiasing: false,
      suppressHiDPIScaling: true,
      maxFps: 60
    });
  }

  initialize() {
    this.currentScene.add(new DebugInfo({ x: -4, y: 0 * SCALE}));
  }
}

var game4 = new Game4();
// game4.toggleDebug();
var loader = new ex.Loader()

loader.addResource(image);
loader.addResource(spriteFontImage);
game4.initialize();
game4.start(loader);
```

**Comment by eonarheim** [comment]
I have on my todo list to update the docs to help users pick the right Text rendering approach for their use case.

We have SDF based text on the roadmap as well which will add a 3rd way of rendering text! So a decision tree becomes important to users

**Comment by eonarheim** [comment]
I've updated some documentation to help folks make a decision about what to use https://excaliburjs.com/docs/text/#font-or-spritefont-or-html

This is fixed in v0.28.0 which was just released!
